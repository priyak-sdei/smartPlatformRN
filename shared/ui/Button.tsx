import React from 'react';
import {
  TouchableOpacity,
  StyleSheet,
  ViewStyle,
  TextStyle,
  TouchableOpacityProps,
} from 'react-native';
import Text from './Text';
import { fonts, moderateScale, spacing, useTheme } from '../theme';

interface ButtonProps extends TouchableOpacityProps {
  title?: string;
  tx?: string;
  txOptions?: import('i18next').TOptions; // Use the TOptions type from i18next
  style?: ViewStyle;
  textStyle?: TextStyle;
  leftIcon?: React.ReactNode;
  rightIcon?: React.ReactNode;
  disabled?: boolean;
}

const Button: React.FC<ButtonProps> = ({
  title,
  tx,
  txOptions,
  style,
  textStyle,
  leftIcon,
  rightIcon,
  disabled = true,
  ...props
}) => {
  const { theme } = useTheme();
  const styles = getStyles(theme);
  return (
    <TouchableOpacity
      onPress={disabled ? undefined : props.onPress}
      style={[styles.button, style, disabled && styles.disabled]}
      disabled={disabled}
      accessibilityState={{ disabled }}
      testID={props.testID || 'button-root'}
      {...props}
    >
      {leftIcon && <>{leftIcon}</>}
      <Text style={[styles.text, textStyle]} tx={tx} txOptions={txOptions}>
        {title}
      </Text>
      {rightIcon && <>{rightIcon}</>}
    </TouchableOpacity>
  );
};

const getStyles = (theme: any) =>
  StyleSheet.create({
    button: {
      flexDirection: 'row',
      alignItems: 'center',
      justifyContent: 'center',
      backgroundColor: theme.colors.primary,
      paddingVertical: moderateScale(spacing.s),
      paddingHorizontal: moderateScale(spacing.l),
      borderRadius: moderateScale(spacing.xxs),
      marginVertical: moderateScale(spacing.xxs),
      width: '100%',
    },
    text: {
      color: theme.colors.background,
      fontFamily: fonts.bold,
      marginHorizontal: moderateScale(spacing.xxs),
    },
    disabled: {
      opacity: 0.6,
    },
  });

export default Button;
