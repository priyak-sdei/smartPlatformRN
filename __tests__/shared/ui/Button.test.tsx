import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import Button from '../../../shared/ui/Button';
import { Text as RNText } from 'react-native';
import renderer, { act } from 'react-test-renderer';

// Mock Text to use RNText so getByText works
jest.mock('../../../shared/ui/Text', () => {
    const { Text } = require('react-native');
    return (props: any) => <Text>{props.children || props.tx}</Text>;
});

const mockTheme = {
    colors: {
        primary: '#d92029',
        background: '#fff',
        disabled: '#ccc',
    },
};

jest.mock('../../../shared/theme', () => ({
    useTheme: () => ({ theme: mockTheme }),
    moderateScale: (v: number) => v,
    spacing: { s: 1, l: 2, xxs: 0.5 },
    fonts: { bold: 'bold' },
}));

describe('Button', () => {
    it('renders with title', () => {
        const { getByText } = render(<Button title="Click me" disabled={false} />);
        expect(getByText('Click me')).toBeTruthy();
    });

    it('renders with tx (i18n)', () => {
        const { getByText } = render(<Button tx="button.tx" disabled={false} />);
        expect(getByText('button.tx')).toBeTruthy();
    });

    it('calls onPress when pressed', () => {
        const onPress = jest.fn();
        const { getByTestId } = render(<Button title="Press" onPress={onPress} disabled={false} />);
        fireEvent.press(getByTestId('button-root'));
        expect(onPress).toHaveBeenCalled();
    });

    it('applies disabled style when disabled', () => {
        const onPress = jest.fn();
        const { getByTestId } = render(<Button title="Disabled" onPress={onPress} disabled />);
        const button = getByTestId('button-root');
        expect(button.props.style.opacity).toBe(0.6);
    });

    it('renders left and right icons', () => {
        const Left = () => <RNText>L</RNText>;
        const Right = () => <RNText>R</RNText>;
        const { getByText } = render(
            <Button title="IconBtn" leftIcon={<Left />} rightIcon={<Right />} disabled={false} />
        );
        expect(getByText('L')).toBeTruthy();
        expect(getByText('R')).toBeTruthy();
    });

    it('applies custom style and textStyle', () => {
        const { getByText } = render(
            <Button title="Styled" style={{ backgroundColor: 'red' }} textStyle={{ color: 'blue' }} disabled={false} />
        );
        expect(getByText('Styled')).toBeTruthy();
    });

    it('matches snapshot', () => {
        let tree: renderer.ReactTestRenderer | null = null;
        act(() => {
            tree = renderer.create(<Button title="Snapshot" disabled={false} />);
        });
        expect(tree!.toJSON()).toMatchSnapshot('Button component');
    });
}); 