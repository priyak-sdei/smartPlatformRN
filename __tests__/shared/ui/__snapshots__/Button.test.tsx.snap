// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button matches snapshot: Button component 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": false,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#d92029",
      "borderRadius": 0.5,
      "flexDirection": "row",
      "justifyContent": "center",
      "marginVertical": 0.5,
      "opacity": 1,
      "paddingHorizontal": 2,
      "paddingVertical": 1,
      "width": "100%",
    }
  }
  testID="button-root"
>
  <Text>
    Snapshot
  </Text>
</View>
`;
